### ISSUE TITLE
Select all shouldn't trigger onRowSelected event for each row

### ISSUE THREAD

**Issue by guido-visser** [ag-grid/ag-grid#3401]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[] support request => see 'Requesting Community Support'
```

**Customers of ag-Grid**

If you are a customer you are entitled to use the ag-Grid's customer support system (powered by Zendesk). Please use that channel for guaranteed response from the ag-Grid team with regards bugs, feature requests and support.

**Requesting Community Support**

If you are not a customer of ag-Grid, ag-grid staff will label your issue as managed-by-the-community. This means that ag-Grid staff is not going to be actively looking into it and it will get closed if inactive for more than one month. The community is welcome to help with this question/support issue.

**Providing a Reproducible Scenario**
Accepted reproducible scenarios are
- A description of the detailed steps to reproduce your behaviour in one of our examples in the docs.
- A plunker

If you decide to send us a plunkr, from any example in our website use the plunkr button in there to fork your own code by following the steps below:

- Select the framework that is appropriate to you from the drop-down
- Open it in plunker. (Use the button plunker in our example)
- Add your changes so that the behaviour is reproduced
- Save and Freeze the plunker(On the top left corner)
- Send us the link to the plunker(You can copy the URL from the browser)

If reporting a bug make sure to state.

Current behaviour.
Expected behaviour. If possible back this up with our docs/examples if possible

**Current behavior**
If you use the select all checkbox in the header, the `onRowSelected` event is triggered for every row. This also happens when using the `this.gridApi.selectAll()`. This results in a freezing browser window when there's a lot of records in it. 

Plnkr to reproduce: https://plnkr.co/edit/7cW6mnVtpxoYOzkKsCEO?p=preview

**Expected behavior**
When using the `this.gridApi.selectAll()` or the select all checkbox in the header, it shouldn't trigger the `onRowSelected` event for every row, but maybe just once with a property provided to differentiate a single row click with a `selectAll` trigger. 

And while I'm on the subject, it would be amazing if a parameter could be added to the `node.setSelected` to trigger the `onRowSelected` event or not, to give more control over when this event is triggered.

**Please tell us about your environment:**

* **ag-Grid version:** 
21.2.1

* **Browser:**
"Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/76.0.3809.132 Safari/537.36"

* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
All

**Comment by davis4acca** [comment]
Hi,

ag-grid staff does not provide support through github. If you are a customer please contact us through our official support channel. Note that this issue has been flagged as managed-by-the-community, we also recommend you to check stackoverflow, note that if there is no activity on this issue after a while we will close it.

Note: You can use a debounce for your action.


Thanks

**Comment by guidogast** [comment]
Thanks for the response, but using a debounce on the function is like: 
![Duct tape fix](http://giphygifs.s3.amazonaws.com/media/Lvand4cUuA6xG/giphy.gif)

I see this as a bug since the product really breaks when you have a lot of records. It shouldn't be triggering the callback for all records. This is highly inefficient.

**Comment by harveyconnor** [comment]
@guido-visser Hi did you ever find a solution to this problem? If so could you please share your resolution. Thanks

**Comment by guido-visser** [comment]
@harveyconnor We have worked around the issue, but I can't share that unfortunately. I still think this is really inefficient of AG-Grid

**Comment by mwood23** [comment]
We can put 100,000 rows of data in there but if we user goes to select all then 100,000 callbacks melt the browser 😦 . For a library that touts performance, I think this should be reconsidered.

**Comment by mwood23** [comment]
Here's a workaround for anyone interested:

```tsx
export function debounce<T extends unknown[], U>(
  callback: (...args: T) => PromiseLike<U> | U,
  wait: number,
) {
  let timer: NodeJS.Timer

  return (...args: T): Promise<U> => {
    clearTimeout(timer)
    return new Promise((resolve) => {
      timer = setTimeout(() => resolve(callback(...args)), wait)
    })
  }
}

// ...

const Foo = () => {
  const debouncedRowSelect = useRef(
    debounce<RowSelectedEvent<ProductItem>[], void>(async (e) => {
      setSelectedRows(e.api.getSelectedNodes())
    }, 300),
  )

  return <AgGridReact onRowSelected={debouncedRowSelect.current}></AgGridReact>
}

```

**Comment by StephenCooper** [comment]
Could I suggest that if there are a lot of rows and selecting all of them is a feature that you use the `onSelectionChanged` [event](https://ag-grid.com/javascript-data-grid/row-selection/#selection-events) instead? This is only fired once, and you can obtain all the currently selected rows from that event via the api.

This would then avoid the need for the workaround and the performance implications. 

Is there a use case where you find that you cannot use `onSelectionChanged` and instead are forced to use `onRowSelected`?

**Comment by theiliad** [comment]
yeah I'd also think `onSelectionChanged` would be used... you could conditionally bail out of `onRowSelected`, but yes indeed that seems like a duct-tape solution.

What would've been great is if there were `onSelectAll`, `onDeselectAll` & `onRowSelect` callbacks

**Comment by edrdesigner** [comment]
You can use `onSelectionChanged` and inside your function you can obtain the selected nodes using `api.forEachNode`
api is from onSelectionChanged(params) 

Example:

```javascript

    api.forEachNode((node) => {
          const nodeRowId = node?.data?.id;

          if (nodeRowId) {
              if (node.isSelected()) {
                  // code 
              } else {
                  // code
              }
          }
      });
```
